### PR TITLE
'dotnet run' doesn't work on self-contained apps

### LINKED ISSUE #527: dotnet build with runtime specified does not create the right output
Moved from https://github.com/dotnet/cli/issues/5039 on behalf of @Fabi, @nguerrera 

----

## Steps to reproduce
1. Install Visual Studio 2017 RC
2. Follow the "Deploying a simple self-contained deployment" part from https://docs.microsoft.com/en-us/dotnet/articles/core/preview3/deploying/ until step 5

## Expected  behavior
After running the dotnet build -r win10-x64 command a folder with the build files (including an .exe file not only dll) should be created at ".\bin\Debug\netcoreapp1.0\<runtime_identifier>"

## Actual behavior
There is no runtime folder with executables generated. That only works with the publish command now.
Before on json based project files it worked fine.

## Environment data
`dotnet --info` output:

```
.NET Command Line Tools (1.0.0-preview4-004233)

Product Information:
 Version:            1.0.0-preview4-004233
 Commit SHA-1 hash:  8cec61c6f7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.0-preview4-004233
```

-----

@nguerrera this is the standalone publish scenario. It should work like this:

## on build
Build produces the portable output as well as a RID-specific output. The RID-specific output contains the portable output + the appropriate RID-specific host for the app. That host still uses deps.json, etc. to load Shared Framework artifacts from the nuget cache

## on publish
The publish output is RID specific. It includes everything from the RID-specific build output + everything from the RID-specific Shared Framework. The publish output can now be zip'd and then deployed to any host with a matching RID.

### PR DESCRIPTION
**Pull request by eerhardt**
Fixing the build output of self-contained applications to be runnable.  Also making 'dotnet run' and Visual Studio F5 work on self-contained applications.

This change also adds an option to append the $(RuntimeIdentifier) to the build output, since the output changes between RIDs.  This would be a breaking change, so the option is defaulted to "off".

@davkean - can you check the usage of the `<None>` item.  It is being used because `<ReferenceCopyLocalPaths>` won't let me rename the file (which is needed for the host executable).  These items shouldn't show up in VS since they are added by a Target, right?

Fix #527

/cc @srivatsn 

## Scenario

When a user sets the `<RuntimeIdentifier>` property in their project, or does a build with `dotnet build -r RID`, the build output is not runnable.  When calling `dotnet run` they get errors like:

```
A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'F:\DotNetTest\DependencyTest\bin\Debug\netcoreapp2.0\'.
```

## Bugs

#527 #791 

## Workarounds

Publish the self-contained app, and then you can run it.

## Risk

~The change to the OutputPath is risky this late.  But @nguerrera said he needed this change for some desktop work as well.  This change makes the output consistent with the output for project.json projects when a RuntimeIdentifier was set.~

By default, there is no change to the OutputPath now.

This change only affects self-contained .NET Core apps during build/run and allows run to work.

## Performance Impact

Minimal, a few more files are being copied to the output folder during build of a self-contained app.

## Regression Analysis

This was a regression from project.json-based projects.

### CODE REVIEW

**Comment by nguerrera** [comment]
For the output path, I don't quite need it, but there are scenarios that are busted without it. dotnet build -r will clobber and I don't think incremental build will do the right thing but I have to check. Here's what I'm thinking, we add 2 knobs: AppendTargetFrameworkToOutputPath and AppendRuntimeIdentifierToOutputPath with respective true, false defaults matching current behavior. Then you can at least opt in if you hit a scenario that is busted and folks have asked repeatedly to opt out of target framework in single-targeting case (in particular, it breaks MVC 5 app built using SDK).

**Review by dsplaisted — APPROVED** [review]
LGTM.

I'd like to see comments explaining what the expected build output is for a self-contained app, and how it differs from the publish output.  Inspecting the code, my understanding is that building a self-contained app should produce an executable in the output path that is directly runnable (without calling `dotnet`), but that still loads its NuGet dependencies from the NuGet package cache.  When you publish it, you will also get all of the dependencies from NuGet packages copied locally.

As for the OutputPath, I agree we should default to including the RuntimeIdentifier in the OutputPath if there is a RuntimeIdentifier specified.

EDIT: Though for my opinion on the OutputPath, I'm not really taking into consideration the potential breaking impact of changing it this late.

**Comment by eerhardt** [comment]
I've responded to the feedback.  I agree adding the RID folder to the OutputPath is too risky at this point.  I've hidden this behavior behind a switch that is defaulted to 'false', as suggested.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets [hunk 1]
```diff
@@ -64,7 +64,10 @@ Copyright (c) .NET Foundation. All rights reserved.
   <!-- Set PublishDir here, before Microsoft.Common.targets, to avoid a competing default there. -->
   <PropertyGroup>
     <PublishDirName Condition="'$(PublishDirName)' == ''">publish</PublishDirName>
-    <PublishDir Condition="'$(PublishDir)' == '' and '$(RuntimeIdentifier)' != ''">$(OutputPath)$(RuntimeIdentifier)\$(PublishDirName)\</PublishDir>
+    <!-- ensure the PublishDir is RID specific-->
+    <PublishDir Condition="'$(PublishDir)' == '' and
+                           '$(AppendRuntimeIdentifierToOutputPath)' != 'true' and
```
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: Should this be '$(AppendRuntimeIdentifierToOutputPath)' == 'true'?

Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: If AppendRuntimeIdentifierToOutputPath is true, the RID will already be in the OutputPath. So we don't need to put the RID in again for the PublishDir.

**Review by nguerrera — APPROVED** [review]

**Comment by nguerrera** [comment]
Thanks, @eerhardt. LGTM.

**Comment by srivatsn** [comment]
Tagging @MattGertz for approval.

**Comment by srivatsn** [comment]
JoC approved offline.

**Pull request merged**